### PR TITLE
docs: Add a code PR advice document

### PR DESCRIPTION
**Pull request by jodh-intel**
Add a document giving advice to code PR authors.

Fixes: #3099.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by fidencio** [comment]
@jodh-intel, I appreciate the document and you have my `lgtm`!
I am **not** approving it yet as I do want to let people go through this, rather than having it merged ASAP.

So, thumbs-up and virtual approval from my side.

**Comment by jodh-intel** [comment]
@fidencio - Thanks - I agree with that approach. We don't want to "Robocop" this PR [1] so let's give folks a few days to digest and comment...

Conversely, silence is compliance... :smile:

---

[1] - *"You have 20 seconds to comply!"* :robot: :gun: :gun: :smile:

**Comment by jodh-intel** [comment]
Note to reviewers: this doc is *not* a coding standard - it's a set of suggestions only.

We could do with some golang advice. Note to self for tomorrow: add golang suggestion to always create a constructor function: we've seen lots of bugs in golang code where code create a struct and "fills in **some** the blanks" leaving other in an invalid state (oops!) Rust doesn't have that problem thankfully.

**Review by cmaf — COMMENTED** [review]
@jodh-intel, all of the unit test advice is helpful but I'm wondering if it is better suited for https://github.com/kata-containers/tests/blob/main/Unit-Test-Advice.md. I think we should definitely emphasize how important it is to write unit tests in this document, but it might be helpful to keep most of the unit test advice in one location.

**Comment by jodh-intel** [comment]
@cmaf - Agreed. In fact I've moved the UT advice document across to this repo since, strictly, unit tests belong "here" whereas the `tests` repo is for other types of tests.

Once this PR lands, I'll remove https://github.com/kata-containers/tests/blob/main/Unit-Test-Advice.md.

**Comment by jodh-intel** [comment]
Ping y'all - please could any passing developers tal at this... :smile:

**Comment by jodh-intel** [comment]
@wainersm - this new document focuses on go+rust. wdyt about shell advice? I could add that into this document, but it *may* be better to have a separate document for that? Or is it even required? There is a lot that could be said about shell and it's an important language so I almost feel it deserves a separate doc if we want to provide shell scripting advice but open to opinions.

**Comment by wainersm** [comment]
@jodh-intel I am going to review it still today :)

File: docs/Unit-Test-Advice.md
Comment (wainersm) [inline docs/Unit-Test-Advice.md]: Should read 'Running' (lead caps) instead of 'running'? Likewise below.

Comment (wainersm) [reply docs/Unit-Test-Advice.md]: Disregard this comment, you changed that line later.

**Review by cmaf — APPROVED** [review]
@jodh-intel lgtm

File: docs/code-pr-advice.md
Comment (bpradipt) [inline docs/code-pr-advice.md:33]: nit: ready -> read

**Comment by bpradipt** [comment]
Really a nice doc @jodh-intel and thanks for putting it together.

**Review by bpradipt — APPROVED** [review]
lgtm

File: docs/code-pr-advice.md
Comment (wainersm) [inline docs/code-pr-advice.md:51]: Can we have an example of copyright/license statement here?

File: docs/code-pr-advice.md
Comment (wainersm) [inline docs/code-pr-advice.md:93]: Do we prefer calling it Go or Golang?

File: docs/code-pr-advice.md
Comment (jodh-intel) [reply docs/code-pr-advice.md]: Good catch @bpradipt! :smile: Updated.

File: docs/code-pr-advice.md
Comment (jodh-intel) [reply docs/code-pr-advice.md:93]: Golang. Even though this is not the standard name, it's a common alias - see the commit message for c1111a1d2d7720cbf0087b384eeb29d5c3fee31c for more detail.

**Comment by wainersm** [comment]
Hi @jodh-intel ,

> @wainersm - this new document focuses on go+rust. wdyt about shell advice? I could add that into this document, but it _may_ be better to have a separate document for that? Or is it even required? There is a lot that could be said about shell and it's an important language so I almost feel it deserves a separate doc if we want to provide shell scripting advice but open to opinions.

I agree advises for shell scripts are important as we have many on kata-containers (6.7%) and tests (60.1%) repositories. I like the idea of having a separate doc.

Things that could be covered:
 -  Quoting variables
 -  Set errexit / nounset /pipefail / errtrace
 -  Use die() and warn() from lib.sh
 - Send error messages to stderr
 - Structure the code around the pattern:
   ```
   var="bar"
   foo() {
   }
   main() {
     foo
   }
   main $@
   ```
 - Use `trap remove_tmp_files EXIT` pattern to clean up resources
 - Default values to variables subject to overwritten (`FOO="${FOO:-foo}"`)
 - Use `local` for local variables
 - Avoid complex logical operations (`foo || bar && foo1 && foo2 || echo "just to print hello World"`)
 - Recommended way to figure out the script's directory

There should be more 10^n (n > 1) things to advice :)

**Comment by jodh-intel** [comment]
@wainersm - do you want to tal at adding such a doc? alternatively, we could create a wiki page, hack for a few days and then create a doc from that?

**Pull request merged**
